lib/elixir/unicode/security.ex

```diff
@@ -108,4 +124,252 @@ defmodule String.Tokenizer.Security do
+      index = Map.get(lookup, script_name)
+      add_script_index_to_scriptset.(scripts_acc_bitmap, index)
+    end)
+  end
+
+  # we can build list of codepoint-to-scriptset mappings
+  # from a file and a scriptname lookup
+  find_matches =
+    &(Regex.run(&2, &1, capture: :all_but_first) || Regex.run(&3, &1, capture: :all_but_first))
+
+  build_scriptsets = fn fname, lu ->
+    single = ~r/^ *([0-9A-F]+) *; *([^#]+) *#/u
+    range = ~r/^ *([0-9A-F]+)\.\.([0-9A-F]+) *; *([^#]+) *#/u
+    matches =
+      Path.join(__DIR__, fname)
+      |> File.read!()
+      |> String.split(["\r\n", "\n"], trim: true)
+      |> Enum.map(&find_matches.(&1, range, single))
+
+    for ls <- matches, ls != nil, reduce: [] do
+      acc ->
+        case ls do
+          [a, b, scripts] ->
+            entry = {String.to_integer(a, 16), String.to_integer(b, 16), scriptset.(lu, scripts)}
+            [entry | acc]
+
+          [a, scripts] ->
+            [{String.to_integer(a, 16), scriptset.(lu, scripts)} | acc]
+        end
+    end
+  end
+
+  script_extensions_ss = build_scriptsets.("ScriptExtensions.txt", script_shortname_indices)
+  scripts_ss = build_scriptsets.("Scripts.txt", script_indices)
+
+  def scripts(c) when not is_integer(c), do: {:error, :non_integer}
+
+  # precedence is ScriptExtensions.txt first, then Scripts.txt,
+  # then fall back to 'Unknown', per headers of those text files.
+  for entry <- script_extensions_ss ++ scripts_ss do
+    case entry do
+      {a, b, ss} -> def scripts(c) when c in unquote(a)..unquote(b), do: unquote(ss)
+      {a, ss} -> def scripts(unquote(a)), do: unquote(ss)
+    end
+  end
+
+  def scripts(_other) do
+    unquote(scriptset.(script_indices, "Unknown"))
+  end
+
+  # 'resolved script set ignores chars w/Extensions {Common}, {Inherited} and augments
+  #  characters with CJK scripts with their respective writing systems. [ALL] are
+  #  ignored when testing for differences in script."
+  common = scriptset.(script_indices, "Common")
+  inh = scriptset.(script_indices, "Inherited")
+  all_ss = union(inh, common)
+  all_sets = [common, inh, all_ss]
+
+  @compile {:inline, all?: 1}
+  defp all?(ss), do: ss in unquote(all_sets)
+
+  # 'define the 'resolved script set' for a string to be the intersection
+  #  of the augmented script sets over all characters in the string'
+  defp resolve([]), do: {@empty, @empty}
+  defp resolve([c | chars]), do: continue(chars, {scripts(c), scripts(c)})
+  defp continue([c | chars], sets), do: continue(chars, add(sets, scripts(c)))
+  defp continue([], sets), do: sets
+
+  # - 'ignore chars w/Extensions {Common}, {Inherited} when testing for diffs in script'
+  # - 'don't resolve {ø} for all-common identifiers, like "_123"' (Table 1a)
+  defp add({resolved, seen} = sets, ss) do
+    cond do
+      @empty == seen -> {ss, ss}
+      all?(ss) -> sets
+      not all?(ss) and (all?(seen) or seen == @empty) -> {ss, ss}
```

Comment:
We should be able to remove the and:

```suggestion
      all?(seen) or seen == @empty -> {ss, ss}
```

and then I think we should be able to remove the or:

```suggestion
      all?(seen) -> {ss, ss}
```

My concern however is that there is too much logic in the module body that we cannot test directly. My suggestion is for us to expose the `resolve` function so we can unit test. I will have more thoughts shortly. :)